refactor: analytics into atomic operations

brunocorrea23 opened the pull request:
targets #294

iuioiua reviewed (changes requested):
Great! Thanks for this. Please also check that their respective functions have increased the analytics data. E.g. that `users_count` has been incremented in the `createUser()` test.

Can we also investigate removing `incrementAnalyticsMetricPerDay()`?

brunocorrea23 commented:
Thanks!
- I have added tests for createUser() and createItem(). I didn't do it for createVote() since this test has not been created yet and is part of an ongoing PR. We can add that later.
- I've removed the incrementVisitsPerDay() in favor of a more generic incrementAnalyticsMetricPerDay() that can be reused if more cases in the future. Also did some refactoring around that. Let me know if this makes sense. I can undo these changes if it doens't.

iuioiua commented on utils/db.ts:
```diff
@@ -401,25 +400,24 @@ export async function incrementAnalyticsMetricPerDay(
     .commit();
 }
 
-export async function incrementVisitsPerDay(date: Date) {
-  // convert to ISO format that is zero UTC offset
-  const visitsKey = [
-    "visits",
-    `${date.toISOString().split("T")[0]}`,
-  ];
-  await kv.atomic()
-    .sum(visitsKey, 1n)
-    .commit();
-}
-
```
Niiiice

iuioiua commented on utils/db.ts:
```diff
@@ -308,6 +306,7 @@ export async function createUser(user: User) {
     .set(usersKey, user)
     .set(usersByLoginKey, user)
     .set(usersBySessionKey, user)
+    .sum(analyticsKey("users_count"), 1n)
```
Hm... It might be better to define a `const usersCount` at the start of the function with the others without using `analyticsKey()`. It gives a more immediate idea of what's going on, IMO. At most, maybe a `toDate()` that returns `date.toISOString().split("T")[0]` might be useful. Ditto for other similar scenarios. WDYT?

iuioiua replied:
I.e. check out `formatDate()` in #289.

brunocorrea23 replied:
Makes sense. Will adjust that!

iuioiua commented on utils/db_test.ts:
```diff
@@ -71,7 +72,14 @@ Deno.test("[db] (get/create)Item()", async () => {
 
   assertEquals(await getItem(item.id), null);
 
+  const itemsCount =
+    (await getAnalyticsMetricPerDay("items_count", new Date()))?.valueOf() ??
+      0n;
```
```suggestion
  assertEquals(await getAnalyticsMetricPerDay("items_count", new Date()), null);
```

brunocorrea23 replied:
@iuioiua , can you pls explain this suggestion here?
I believe that if you do it like this, the assertion is going to fail. When it gets to this point, the itemsCount is probably equal to `2n` because of the two `createItem()` that were run on the test above (`" [db] getAllItems()"`)

We can't assume anything about the value o itemsCount, as we don't know how many tests will use it before in the code. We can only test that the itemsCount_after = itemsCount_before + 1.

Does it make sense? Did I understand correctly your suggestion?

iuioiua replied:
Perhaps, we should move the line that defines `itemsCount` elsewhere such that it doesn't require a default value. WDYT?

iuioiua reviewed (changes requested):
IMO, we should drop `getAnalyticsMetricPerDay()` in favour of, say `getItemsCountByDate()`, and the same for the others. I'd rather have more lines of code that are easier to understand rather than fewer lines of code that require more thinking. WDYT?

iuioiua commented on utils/db_test.ts:
```diff
@@ -71,7 +72,14 @@ Deno.test("[db] (get/create)Item()", async () => {
+      0n;
   await createItem(item);
+  assertEquals(
+    (await getAnalyticsMetricPerDay("items_count", new Date()))!.valueOf(),
```
Is `.valueOf()` needed? Ditto elsewhere.

brunocorrea23 replied:
It seems so. This is the test result when I remove the velueOf():

[Diff] Actual / Expected
-   [Deno.KvU64: 3n]
+   3n

iuioiua replied:
I see. Fair enough.

brunocorrea23 commented:
> IMO, we should drop `getAnalyticsMetricPerDay()` in favour of, say `getItemsCountByDate()`, and the same for the others. I'd rather have more lines of code that are easier to understand rather than fewer lines of code that require more thinking. WDYT?

Thanks for your input here 🙏  I'll do some refactoring around this so that we end up with 3 functions per metric:
- increment[X]ByDay(date: Date)
- get[X]ByDay(date: Date)
- get[X]ListByDay(options?: Deno.KvListOptions)

iuioiua commented on utils/db.ts:
```diff
@@ -69,6 +74,7 @@ export async function createItem(item: Item) {
   const itemsKey = ["items", item.id];
   const itemsByTimeKey = ["items_by_time", item.createdAt.getTime(), item.id];
   const itemsByUserKey = ["items_by_user", item.userId, item.id];
+  const itemsCountKey = ["items_count", `${formatDate(new Date())}`];
```
```suggestion
  const itemsCountKey = ["items_count", formatDate(new Date())];
```

iuioiua commented on utils/db.ts:
```diff
@@ -179,6 +185,7 @@ export async function createVote(vote: Vote) {
     vote.item.id,
     vote.user.id,
   ];
+  const votesCountKey = ["votes_count", `${formatDate(new Date())}`];
```
```suggestion
  const votesCountKey = ["votes_count", formatDate(new Date())];
```

brunocorrea23 commented:
@iuioiua , couple of comments here:
- I've created the analytics functions I've mentioned above.
- I haven't changed the way /stats uses the analytics metrics as I believe this is outside the scope of this PR and might overlap with the work on #289 . If the changes are not handled by that PR, I can create another one soon after it.
- As for the default value on the db_test.ts you've mentioned above: the way I see to remove the need for a default value is simply to move the function to the top of the file, where we can guarantee that no one has increased that metrics count before. But I do think that the current implementation is more robust, as it doesn't assume anything about order of functions within that file. I can imagine in the future someone trying to changing that file, the test starts failing and it will be hard to debug why.

iuioiua commented on utils/db.ts:
```diff
@@ -387,39 +394,90 @@ export function compareScore(a: Item, b: Item) {
-  metric: string,
-  date: Date,
-) {
+export async function incrementVisitsCountByDay(date: Date) {
```
```suggestion
export async function incrVisitsCountByDay(date: Date) {
```
Nit. WDYT? Is it still clear to a newcomer what this does?

brunocorrea23 replied:
Makes sense. I think it is still clear!

iuioiua commented on utils/db.ts:
```diff
@@ -387,39 +394,90 @@ export function compareScore(a: Item, b: Item) {
+export async function getVotesCountByDay(date: Date) {
+  return await getValue<bigint>([
+    "votes_count",
+    `${date.toISOString().split("T")[0]}`,
```
```suggestion
    formatDate(date),
```

iuioiua commented on utils/db.ts:
```diff
@@ -387,39 +394,90 @@ export function compareScore(a: Item, b: Item) {
+export async function getUsersCountDay(date: Date) {
+  return await getValue<bigint>([
+    "users_count",
+    `${date.toISOString().split("T")[0]}`,
```
```suggestion
    formatDate(date),
```

iuioiua reviewed (changes requested):
Few nits

brunocorrea23 commented:
Addressed the previous comments. Please take a look.

iuioiua reviewed (approved):
Great work! Nice cleanups and simplifications! Thanks for this one, @brunocorrea23.

iuioiua commented:
Closes #294

brunocorrea23 commented:
@iuioiua , thank you for all the help and input! 🙏
